db.go

```diff
@@ -54,6 +69,91 @@ func (e *Entry) GetAttr(attr string) (Attr, bool) {
 	return v, ok
 }
 
+// Authenticate checks that the LDAP Entry is valid for authentication and that
+// the entry has a password that matches the given password. If the entry is
+// valid for authentication and the password matches the one in the entry, nil
+// is returned. Otherwise an error is returned.
+//
+// The form of hashed password in an entry is described by [RFC 2307, 5.3]
+// Interpreting user and group entries with the supported schemes listed in the
+// [hashSchemes] map above. This is how [OpenLDAP passwords] are done, so this
+// does it the same. e.g. "{SSHA}<base64-encoded-hash+salt>"
+//
+// [RFC 2307, 5.3]: https://datatracker.ietf.org/doc/html/rfc2307#section-5.3
+// [OpenLDAP passwords]: https://www.openldap.org/faq/data/cache/347.html
+func (e *Entry) Authenticate(password string) error {
+	if !e.isValidForAuthentication() {
```

Comment:
very nitpicky, but I'm more used to reading:
`if err := validateThisOrThat; err := nil { //....` 
meaning that the validation function is in charge of the error returned, but I can be convinced otherwise.